### PR TITLE
Add host verification to NetClient

### PR DESCRIPTION
**Pull request by dan-lind**
Add host verification to NetClient
Added missing tests for host verification to HttpTLSTest

Please note breaking change in NetSocket.java
public NetSocket upgradeToSsl(Handler<Void> handler);
changes to
public NetSocket upgradeToSsl(String host, int port, Handler<Void> handler);

Signed-off-by: Daniel Lindberg <daniel.lindberg@omegapoint.se>

### CODE REVIEW

File: src/main/java/io/vertx/core/net/NetSocket.java [hunk 1]
```diff
@@ -217,7 +217,7 @@ default NetSocket sendFile(String filename, long offset, Handler<AsyncResult<Voi
    * @return a reference to this, so the API can be used fluently
    */
   @Fluent
-  NetSocket upgradeToSsl(Handler<Void> handler);
+  NetSocket upgradeToSsl(String host, int port, Handler<Void> handler);
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/NetSocket.java]: breaking API change + javadoc is not updated

File: src/test/java/io/vertx/test/core/TLSCert.java [hunk 2]
```diff
@@ -164,7 +164,26 @@ TrustOptions getClientTrustOptions() {
     KeyCertOptions getClientKeyCertOptions() {
       return new PemKeyCertOptions().setKeyPath("tls/client-key.pem").setCertPath("tls/client-cert-ca.pem");
     }
-  };
+  },
+
+    FAKE() {
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/TLSCert.java]: the procedure for producing the fake-server-keystore.jks should be documented in _ssl.txt_

**Comment by vietj** [comment]
this features provided for generic TCP clients, relies on the HTTPS rules for host verification, that is a problem.

**Comment by vietj** [comment]
I think this is the main problem with this feature, is that it is only for HTTPS rules. LDAP has other rules for instance. If someone would be an LDAP client on top of NetClient he would use the LDAP rules which are different, etc...

**Comment by dan-lind** [comment]
Hmm, yeah I see your point. Looking at the sun.security.ssl.X509TrustManagerImpl.checkIdentity() it looks like the only two available algoritms are HTTPS and LDAP/LDAPS. But I guess we currently have no idea of knowing if a NetClient is an LDAP client or not?

**Comment by vietj** [comment]
beyond that the client might have its own way for validating the certificate.

**Comment by dan-lind** [comment]
Any thoughts on how to work around that? It would be nice if we could provide an OOTB way to use the HTTPS rules for host verification if you are using a plain TCP client rather than LDAP or whatnot. As it stands now you would basically have to fork master and apply something along the lines of this PR.

**Comment by dan-lind** [comment]
Perhaps a simple solution would be to disable it by default (rather than enable like the HTTP client). 
Then rename the method to something like NetClientOptions.setVerifyHostUsingHttpsRules() and clearly state in the documentation that the hostname will be verified according to HTTPS rules.

**Comment by vietj** [comment]
I would say that in SSLHelper the boolean verifyHost would be changed to a String that is the identifier of the verification algorithm.

HttpClientOption would set it to HTTPS if verifyHost == true

NetClientOptions would have then a property String hostnameVerifier that could be either HTTPS or LDAPS .

**Comment by dan-lind** [comment]
I can fix that. Only problem is that I don't really know how to test the LDAP verification...

**Comment by vietj** [comment]
do it this way, test that "HTTPS" string works, it's a feature of the SSLEngine anyway, so we would test this feature actually. The important point is that we allow the user to configure verification in general and that it works for HTTPS that is an important use case.

**Comment by dan-lind** [comment]
Besides updating the javadoc for NetSocket, should the breaking change be documented in some special way?

**Comment by vietj** [comment]
there should be no breaking changes

**Comment by dan-lind** [comment]
Probably need to add an overloaded upgradeToSsl then, the current method breaks if you enable host verification since it doesn't pass the hostname to the ssl engine.

**Comment by vietj** [comment]
the current should reuse the host/port specified when the NetSocket is created instead

**Comment by dan-lind** [comment]
I think I have made all the necessary changes now, please review

**Comment by vietj** [comment]
yup looks good to me 👍

**Comment by vietj** [comment]
have you signed the eclipse CLA @dan-lind ?

**Comment by dan-lind** [comment]
Yep

File: src/main/java/io/vertx/core/net/impl/SSLHelper.java [hunk 3]
```diff
@@ -410,9 +407,9 @@ private SslHandler createHandler(SSLEngine engine, boolean client) {
           break;
         }
       }
-    } else if (verifyHost) {
+    } else if (!endpointIdentificationAlgorithm.isEmpty()) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/impl/SSLHelper.java]: endpointIdentificationAlgorithm could be null with a malformed option value in NetClientOptions : add a non null ckeck too

Comment (dan-lind) [inline src/main/java/io/vertx/core/net/impl/SSLHelper.java]: Perhaps endpointIdentificationAlgorithm should be an enum instead? Would remove both null and invalid option values.

Comment (vietj) [inline src/main/java/io/vertx/core/net/impl/SSLHelper.java]: no need to add an enum for this

Comment (dan-lind) [inline src/main/java/io/vertx/core/net/impl/SSLHelper.java]: Added Objects.requireNonNull for the setter, I think this is preferred over null checking everywhere the value might be used.

**Comment by vietj** [comment]
thanks for the contribution!

**Pull request merged**